### PR TITLE
[Enhancement] Do rocksdb compact before data load if load cost a lot of time

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Rocksdb range deletes may cause range query to perform badly and this pr(https://github.com/StarRocks/starrocks/pull/11513) changes delvec GC logic to scan through all delvecs of a tablet, get all expired delvec keys, then perform a batch delete. However, there are still some scenarios where the range query performs poorly and hangs BE process starts, and the start time is significantly reduced after finish compaction. So we will load tablet with a limit time first, if we can't load all tablets in the limit time, we will trigger a meta compaction and then load again.
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

File: be/src/storage/kv_store.cpp [hunk 1]
```diff
@@ -242,17 +243,39 @@ Status KVStore::iterate(ColumnFamilyIndex column_family_index, const std::string
     } else {
         it->Seek(prefix);
     }
-    for (; it->Valid(); it->Next()) {
-        if (!prefix.empty()) {
-            if (!it->key().starts_with(prefix)) {
-                return Status::OK();
+    // if limit time is less than zero, it means no limit
```
Comment (howrocks) [inline be/src/storage/kv_store.cpp:246]: <= 0?

File: be/src/common/config.h [hunk 2]
```diff
@@ -903,4 +903,6 @@ CONF_Int32(binlog_page_max_size, "1048576");
 
 CONF_mInt64(txn_info_history_size, "20000");
 
+CONF_mInt64(load_tablet_limit_time, "30");
```
Comment (howrocks) [inline be/src/common/config.h:906]: load_tablet_limit_seconds is clearer.

**Comment by howrocks** [comment]
This enhancement should be backported to version from 2.3+.

File: be/src/storage/kv_store.cpp (hunk 1 above)
Comment (sevev) [reply be/src/storage/kv_store.cpp]: > <= 0?

done, change in https://github.com/StarRocks/starrocks/pull/19347/commits/ab09d917afdd3617d180b83cb64bf9ae98ec8cd2

File: be/src/common/config.h (hunk 2 above)
Comment (sevev) [reply be/src/common/config.h]: > load_tablet_limit_seconds is clearer.

done, changed in https://github.com/StarRocks/starrocks/pull/19347/commits/ab09d917afdd3617d180b83cb64bf9ae98ec8cd2

File: be/src/storage/data_dir.cpp [hunk 3]
```diff
@@ -298,7 +298,22 @@ Status DataDir::load() {
...
+    Status load_tablet_status =
+            TabletMetaManager::walk_with_limit_time(_kv_store, load_tablet_func, config::load_tablet_limit_seconds);
+    if (load_tablet_status.is_time_out()) {
+        Status s = _kv_store->compact();
+        if (!s.ok()) {
+            LOG(ERROR) << "data dir " << _path << " compact meta befor load failed";
+            return s;
+        }
+        for (auto tablet_id : tablet_ids) {
+            _tablet_manager->drop_tablet(tablet_id, kKeepMetaAndFiles);
```
Comment (trueeyu) [inline be/src/storage/data_dir.cpp:310]: is this operation risky？

Comment (sevev) [reply be/src/storage/data_dir.cpp:310]: > is this operation risky？

This operation is only remove tablet in memory but not remove data file and meta file in disks, so i think this operation is safe.

**Review by trueeyu — DISMISSED** [review]

**Review by mark-phoenix-ai — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/storage/tablet_meta_manager.cpp [hunk 4]
```diff
@@ -428,6 +428,13 @@ Status TabletMetaManager::remove(DataDir* store, TTabletId tablet_id, TSchemaHas
 Status TabletMetaManager::walk(
         KVStore* meta,
         std::function<bool(long /*tablet_id*/, long /*schema_hash*/, std::string_view /*meta*/)> const& func) {
+    return walk_with_limit_time(meta, func, -1);
+}
+
+Status TabletMetaManager::walk_with_limit_time(
```
[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/tablet_meta_manager.cpp]: done

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/tablet_updates.cpp [hunk 5]
```diff
@@ -233,33 +233,67 @@ Status TabletUpdates::_load_from_pb(const TabletUpdatesPB& tablet_updates_pb) {
...
+    RETURN_IF_ERROR(TabletMetaManager::del_vector_iterate(
+            _tablet.data_dir()->get_meta(), _tablet.tablet_id(), min_rsid, max_rsid,
+            [&](uint32_t segment_id, int64_t version, std::string_view value) -> bool {
+                auto iter = del_vector_by_rsid.lower_bound(segment_id);
+                if (iter == del_vector_by_rsid.end()) {
+                    std::string msg = strings::Substitute("found delete vector of non exist rowset");
+                    _set_error(msg);
+                    LOG(ERROR) << msg;
+                    return false;
+                }
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:263]: maybe just print error and ignore in production?

File: be/src/storage/tablet_updates.cpp [hunk 6]
```diff
@@ -233,33 +233,67 @@ Status TabletUpdates::_load_from_pb(const TabletUpdatesPB& tablet_updates_pb) {
...
     // TODO: save num_dels in rowset meta.
+
+    std::map<uint32_t, std::vector<std::pair<int32_t, DelVectorPtr>>> del_vector_by_rsid;
+    uint32_t max_rsid = 0;
+    uint32_t min_rsid = 1 << 31;
+    for (auto& [rsid, rowset] : _rowsets) {
+        if (unapplied_rowsets.find(rsid) == unapplied_rowsets.end()) {
+            std::vector<std::pair<int32_t, DelVectorPtr>> del_vectors;
+            del_vectors.resize(rowset->num_segments());
+            del_vector_by_rsid[rsid] = del_vectors;
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:244]: del_vector_by_rsid[rsid].resize(rowset->num_segments())
and maybe you should make sure version is initialized as 0?

File: be/src/storage/tablet_updates.cpp [hunk 7]
```diff
@@ -233,33 +233,67 @@ Status TabletUpdates::_load_from_pb(const TabletUpdatesPB& tablet_updates_pb) {
...
+                if (iter == del_vector_by_rsid.end()) {
+                    std::string msg = strings::Substitute("found delete vector of non exist rowset");
+                    _set_error(msg);
+                    LOG(ERROR) << msg;
+                    return false;
+                }
+                uint32_t rsid = iter->first;
+                uint32_t idx = segment_id - rsid;
+                DCHECK(idx < iter->second.size());
+                if (version >= iter->second[idx].first) {
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:267]: what's the initial value of `iter->second[idx].first` ? is it 0 for sure?

File: be/src/storage/tablet_updates.cpp (hunk 5 above)
Comment (sevev) [reply be/src/storage/tablet_updates.cpp]: > maybe just print error and ignore in production?

done

File: be/src/storage/tablet_updates.cpp (hunk 6 above)
Comment (sevev) [reply be/src/storage/tablet_updates.cpp]: > del_vector_by_rsid[rsid].resize(rowset->num_segments()) and maybe you should make sure version is initialized as 0?

Good advice, changed in https://github.com/StarRocks/starrocks/pull/19347/commits/7bf7ae3af2f36089321a824401b85d2f42945935

File: be/src/storage/tablet_updates.cpp (hunk 7 above)
Comment (sevev) [reply be/src/storage/tablet_updates.cpp:266]: > what's the initial value of `iter->second[idx].first` ? is it 0 for sure?

Has initialized as 0.

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.3

**Pull request merged**